kerngen/pisa_generators/rescale.py

```diff
@@ -0,0 +1,139 @@
+                range(polys.input_remaining_rns.parts),
+                range(polys.input_remaining_rns.rns),
+                range(self.context.units),
+            )
+        ]
+
+    def add_last_half(self, output, input0, input1, input_remaining_rns):
+        """Add input0 to input1 (first part)"""
+        return self.partial_op(
+            pisa_op_add,
+            self.PartialOpPolys(output, input0, input1, input_remaining_rns),
+            self.PartialOpOptions(
+                output_last_q=True,
+                input0_last_q=True,
+                input1_last_q=True,
+                input1_first_part=True,
+                op_last_q=True,
+            ),
+        )
+
+    def sub_last_half(self, output, input0, input1, input_remaining_rns):
+        """Subtract input1 (first part) with input0 (last RNS)"""
+        return self.partial_op(
+            pisa_op_sub,
+            self.PartialOpPolys(output, input0, input1, input_remaining_rns),
+            self.PartialOpOptions(input0_last_q=True, input1_first_part=True),
+        )
+
+    def to_pisa(self) -> list[PIsaOp]:
+        """Return the p-isa code to perform an mod switch down"""
+        # Convenience and Immediates
+        context = self.context
+        last_q = self.input0.rns - 1
+        one = Immediate(name="one")
+        r2 = Immediate(name="R2", rns=last_q)
+        iq = Immediate(name="iq", rns=last_q)
+        q_last_half = Polys("qLastHalf", 1, self.input0.rns)
+        q_i_last_half = Polys("qiLastHalf", 1, rns=last_q)
+
+        # Drop down input rns
+        input_last_rns = Polys.from_polys(self.input0, mode="last_rns")
+        input_remaining_rns = Polys.from_polys(self.input0, mode="drop_last_rns")
+
+        # Temp.
+        y = Polys(
+            "y",
+            input_last_rns.parts,
+            input_last_rns.rns,
+            start_rns=input_last_rns.start_rns,
+        )
+        x = Polys("x", input_remaining_rns.parts, input_remaining_rns.rns)
+
+        # Compute the `delta_i = t * [-t^-1 * c_i] mod ql` where `i` are the parts
+        # The `one` acts as a select flag as whether or not R2 the Montgomery
+        # factor should be applied
+        return mixed_to_pisa_ops(
+            [
+                Comment("Start of Rescale kernel."),
+                INTT(context, y, input_last_rns),
+                Muli(context, y, y, one),
+                Comment("Add the last part of the input to y"),
+                self.add_last_half(y, y, q_last_half, input_remaining_rns),
+                Comment("Subtract q_i (last half/last rns) from y"),
+                self.sub_last_half(x, y, q_i_last_half, input_remaining_rns),
+                Muli(context, x, x, r2),
+                NTT(context, x, x),
+                Sub(context, x, Polys.from_polys(self.input0, mode="drop_last_rns"), x),
+                Muli(context, self.output, x, iq),
+                Comment("End of Rescale kernel."),
+            ]
+        )
```

Comment:
For reasons of code clarity, code maintainability,  plus since there is quite a lot of commonality between this code and the MOD kernel in [mod.py](./pisa_generators/mod.py), we should try to avoid cutting and pasting  segments of code that do the same ting but in different files.

Reply:
@faberga please review the latest changes. I created few helper methods that are used in both mod and rotate. It eliminates the code replications as you mentioned.